### PR TITLE
New `VotingEntry` type

### PR DESCRIPTION
**Pull request by newhoggy**
# Changelog

```yaml
- description: |
    New `VotingEntry` type
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
  - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - improvement    # QoL changes e.g. refactoring
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

In a future version of `cardano-ledger`, the voter and governance action id have been moved out of voting procedure so we require a type to carry the triplet for serialisation.  `VotingEntry` will be that new type.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] The change log section in the PR description has been filled in
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - round trip tests
  - integration tests
  See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [ ] The changelog section in the PR is updated to describe the change
- [ ] Self-reviewed the diff

### CODE REVIEW

**Review by carbolymer — APPROVED** [review]

File: cardano-api/internal/Cardano/Api/Governance/Actions/VotingEntry.hs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import           Cardano.Api.SerialiseCBOR (FromCBOR (fromCBOR), SerialiseAsCBOR (..),
+                   ToCBOR (toCBOR))
+import           Cardano.Api.SerialiseTextEnvelope
+
+import qualified Cardano.Binary as CBOR
+
+data VotingEntry era = VotingEntry
+  { votingEntryVoter            :: Voter era
+  , votingEntryGovActionId      :: GovernanceActionId era
+  , votingEntryVotingProcedure  :: VotingProcedure era
```
Comment (Jimbo4350) [inline cardano-api/internal/Cardano/Api/Governance/Actions/VotingEntry.hs:23]: Aren't we repeating types? `VotingProcedure` contains the voter and the governance action id:
```
createVotingProcedure
  :: ShelleyBasedEra era
  -> Vote
  -> Voter era
  -> GovernanceActionId era
  -> VotingProcedure era
 ```

**Review by Jimbo4350 — APPROVED** [review]
LGTM, can you add a `TODO: Conway era` comment to the code and copy paste the description in that comment.

File: cardano-api/internal/Cardano/Api/Governance/Actions/VotingEntry.hs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  } deriving (Show, Eq)
+
+instance IsShelleyBasedEra era => ToCBOR (VotingEntry era) where
```
Comment (Jimbo4350) [inline cardano-api/internal/Cardano/Api/Governance/Actions/VotingEntry.hs:26]: Let's rely on the ledger type's serialization instances. We don't want non CDDL serialization as this has bitten us in the past.

Comment (newhoggy) [reply cardano-api/internal/Cardano/Api/Governance/Actions/VotingEntry.hs:30]: Added a comment on the type to wrap a ledger type instead.

**Pull request merged**